### PR TITLE
feat: create reusable player component

### PR DESCRIPTION
**Pull request by sabababluani**

### CODE REVIEW

File: app/Components/AudioPlayer/AudioPlayer.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import styles from './AudioPlayer.module.scss';
+import { useAudioPlayer } from '@/app/Helpers/AudioHelpers';
+import MusicPhoto from './Components/MusicPhoto/MusicPhoto';
+import ArrowLink from './Components/Arrow/Arrow';
+import Player from './Components/Player/Player';
+import { SongProps } from '@/app/Interfaces/Interfaces';
+
+
+const AudioPlayer = ({ songs }: SongProps) => {
```
Comment (pailodze) [inline app/Components/AudioPlayer/AudioPlayer.tsx:10]: do not destrucute is props: SongProps

Comment (pailodze) [inline app/Components/AudioPlayer/AudioPlayer.tsx:10]: Name song props AudioPlayerPropsInterface

File: app/Components/AudioPlayer/AudioPlayer.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        handleProgressChange,
+        PlayMusic,
+        handleTenSecondsBack,
+        handleNextSong,
+        handlePreviousSong,
+        handleVolumeDown,
+        handleVolumeUp,
+    } = useAudioPlayer(songs);
+
+    const isPlaying = audioRef.current ? !audioRef.current.paused : false;
```
Comment (pailodze) [inline app/Components/AudioPlayer/AudioPlayer.tsx:25]: maybe audioRef.current && !audioRef.current.paused should work

File: app/Components/AudioPlayer/Components/Arrow/Arrow.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import Image from "next/image";
+import styles from './Arrow.module.scss';
+import Link from 'next/link';
+
+const ArrowLink = () => (
+        <Link href='#' className={styles.arrow}>
```
Comment (pailodze) [inline app/Components/AudioPlayer/Components/Arrow/Arrow.tsx:6]: What is this link for?

File: app/Components/AudioPlayer/Components/InputRange/InputRange.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import styles from './InputRange.module.scss';
+import { InputRangeProps } from '@/app/Interfaces/Interfaces';
+
+
+const InputRange: React.FC<InputRangeProps> = ({ defaultValue, onChange, progressRef }) => {
+    const [audioPlayer, setAudioPlayer] = useRecoilState(audioPlayerState);
+
+    useEffect(() => {
+        if (progressRef.current) {
+            progressRef.current.value = String((audioPlayer.currentTime / audioPlayer.duration) * 100);
```
Comment (pailodze) [inline app/Components/AudioPlayer/Components/InputRange/InputRange.tsx:13]: extract value in variable with descriptive name

File: app/Components/AudioPlayer/Components/Player/Player.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import AdjustButtons from '../AdjustButtons/AdjustButtons';
+import ProgressBars from '../ProgressBars/ProgressBars';
+import styles from './Player.module.scss';
+import { Player } from '../../../../Interfaces/Interfaces'
+
+const Player: React.FC<Player> = ({ playing, currentTime, duration, progressRef, ipadProgressRef, onProgressChange, onTenSecondsBack, onVolumeDown, onVolumeUp, onPreviousSong, onNextSong, onPlayMusic, }) => {
```
Comment (pailodze) [inline app/Components/AudioPlayer/Components/Player/Player.tsx:7]: same here props: PlayerPropsInterface

File: app/Components/MainAudioPlayer/MainAudioPlayer.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import styles from './MainAudioPlayer.module.scss';
+import { useAudioPlayer } from '@/app/Helpers/AudioHelpers';
+import { SongProps } from '@/app/Interfaces/Interfaces';
+import MainAdjust from './Components/MainAdjust/MainAdjust';
+import ProgressBar from './Components/MainProgress/MainProgress';
+
+const MainAudioPlayer: React.FC<SongProps> = ({ songs }) => {
```
Comment (pailodze) [inline app/Components/MainAudioPlayer/MainAudioPlayer.tsx:8]: I think there exists duplicate code

File: app/Components/MainAudioPlayer/Components/MainProgress/MainProgress.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                defaultValue={String((currentTime / duration) * 100)}
+                onChange={handleProgressChange}
+                progressRef={progressRef}
+            />
+            <div className={styles.mobileTime}>
+                <span>{`${formatTime(currentTime)}/${formatTime(duration)}`}</span>
+            </div>
+            <p>{formatTime(duration)}</p>
+            <div className={styles.icons} onClick={toggleLoop}>
+                <Image src={loop ? "/greenLoop.png" : "/disableLoop.png"} alt='loop'  width={28} height={28}/>
```
Comment (pailodze) [inline app/Components/MainAudioPlayer/Components/MainProgress/MainProgress.tsx:32]: extract this checking as variable with descriptive name

File: app/Helpers/AudioHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        const handleEnded = () => {
+            setAudioPlayer((prev) => ({
+                ...prev,
+                currentSongIndex: (prev.currentSongIndex + 1) % songs.length,
+                currentTime: 0,
+            }));
+        };
+
+        if (audioRef.current) {
+            const audio = audioRef.current;
```
Comment (pailodze) [inline app/Helpers/AudioHelpers.ts:37]: extract this to function with descriptive name

File: app/Helpers/AudioHelpers.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+    const handleProgressChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        if (audioRef.current) {
+            const newTime = (e.target.valueAsNumber / 100) * audioPlayer.duration;
+            audioRef.current.currentTime = newTime;
+            setAudioPlayer((prev) => ({ ...prev, currentTime: newTime }));
+        }
+    };
+
+    const PlayMusic = () => {
```
Comment (pailodze) [inline app/Helpers/AudioHelpers.ts:77]: function name should be in camelCase

File: app/Helpers/AudioHelpers.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
+import { useRef, useEffect } from 'react';
```
Comment (pailodze) [inline app/Helpers/AudioHelpers.ts:1]: move this audiohelpers to AudioPlayer/hooks/use-audio.hook.ts

File: app/Interfaces/Interfaces.ts [hunk 11]
```diff
@@ -8,3 +8,91 @@ export interface Button {
     icon?: string;
     type?: ButtonType;
 }
+
+export interface MusicPhotoProps {
```
Comment (pailodze) [inline app/Interfaces/Interfaces.ts:12]: name this all interface in its own files

File: app/array.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+const songs = [
```
Comment (pailodze) [inline app/array.ts:1]: move this fiel to AudioPlayer/utils/dummy-musics.ts

**Review by pailodze — CHANGES_REQUESTED** [review]

File: app/Components/AudioPlayer/Components/Arrow/Arrow.tsx (hunk 3 above)
Comment (sabababluani) [reply app/Components/AudioPlayer/Components/Arrow/Arrow.tsx:6]: music page isnot ready for now it is not permanent

File: app/Components/BigPlayer/BigPlayer.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  return (
+    <>
+      <audio
+        ref={audioRef}
+        src={props.songs[audioPlayer.currentSongIndex].audioSrc}
+        loop={audioPlayer.loop}
+      ></audio>
+      <div className={styles.container}>
+        <div className={styles.controls}>
+          <BigPlayerAdjust
```
Comment (pailodze) [inline app/Components/BigPlayer/BigPlayer.tsx:33]: Rename MediumPlayerTools

File: app/Components/BigPlayer/BigPlayer.tsx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+            loop={audioPlayer.loop}
+            toggleLoop={toggleLoop}
+          />
+        </div>
+      </div>
+    </>
+  );
+};
+
+export default BigPlayer;
```
Comment (pailodze) [inline app/Components/BigPlayer/BigPlayer.tsx:55]: name it MediumPlayer

File: app/Components/BigPlayer/Components/BigPlayerAdjust/BigPlayerAdjust.tsx [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import Image from "next/image";
+import React from "react";
+import { MainAdjustProps } from "../../Interfaces/BigAudioPlayer-props.interface";
+import styles from "./BigPlayerAdjust.module.scss";
+
+const BigPlayerAdjust = (props: MainAdjustProps) => {
```
Comment (pailodze) [inline app/Components/BigPlayer/Components/BigPlayerAdjust/BigPlayerAdjust.tsx:6]: name this MediumPlayerTools

File: app/Components/SmallPlayer/Components/Player/Player.tsx [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import React from "react";
+import { PlayerProps } from "../../Interfaces/AudioPlayer-props.interface";
+import AdjustButtons from "../AdjustButtons/AdjustButtons";
+import ProgressBars from "../ProgressBars/ProgressBars";
+import styles from "./Player.module.scss";
+
+const Player = (props: PlayerProps) => {
+  return (
+    <div className={styles.musicMiddle}>
+      <AdjustButtons
```
Comment (pailodze) [inline app/Components/SmallPlayer/Components/Player/Player.tsx:10]: Tools

File: app/Components/SmallPlayer/Interfaces/AudioPlayer-props.interface.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
+export interface ProgressBarsProps {
+  currentTime: number;
```
Comment (pailodze) [inline app/Components/SmallPlayer/Interfaces/AudioPlayer-props.interface.ts:2]: audio-player-props.interface.ts

**Review by pailodze — CHANGES_REQUESTED** [review]

File: app/Atoms/states.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
+import { atom } from 'recoil';
+
+export const audioPlayerState = atom({
```
Comment (pailodze) [inline app/Atoms/states.ts:3]: atom has not generic type

File: app/Components/MediumPlayer/Components/Shuffle/Shuffle.tsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import { ShuffleIconProps } from './interfaces/Shuffle-Icon-Props.interface';
+
+const ShuffleIcon: React.FC<ShuffleIconProps> = ({ color }) => (
```
Comment (pailodze) [inline app/Components/MediumPlayer/Components/Shuffle/Shuffle.tsx:4]: remove this component

File: app/Components/MediumPlayer/MediumPlayer.tsx [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    handleProgressChange,
+    playMusic,
+    toggleLoop,
+    handleNextSong,
+    handlePreviousSong,
+    handleVolumeDown,
+    handleVolumeUp,
+  } = useAudioPlayer(props.songs);
+
+  const isPlaying = audioRef.current ? audioRef.current.paused : false;
```
Comment (pailodze) [inline app/Components/MediumPlayer/MediumPlayer.tsx:22]: isPlaying = !!audioRef.current.paused

File: app/Components/SmallPlayer/Components/PlayerMediaControl/PlayerMediaControl.tsx [hunk 21]
```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import Image from 'next/image';
+import ProgressBar from '../ProgressBar/ProgressBar';
+import styles from './PlayerMediaControl.module.scss';
+import { formatTime } from '@/app/Helpers/AudioHelpers';
+import { PlayerMediaControlProps } from '../../interfaces/Player-Media-Control-Props.interface';
+
+const PlayerMediaControl = (props: PlayerMediaControlProps) => {
```
Comment (pailodze) [inline app/Components/SmallPlayer/Components/PlayerMediaControl/PlayerMediaControl.tsx:8]: PlayerMediaControlPropsInterface

File: app/Components/SmallPlayer/Components/PlayerTools/PlayerTools.tsx [hunk 22]
```diff
@@ -0,0 +1,57 @@
+import Image from 'next/image';
+import React from 'react';
+import styles from './PlayerTools.module.scss';
+import { AdjustButtonProps } from '@/app/Interfaces/Interfaces';
+
+const PlayerTools = (props: AdjustButtonProps) => {
```
Comment (pailodze) [inline app/Components/SmallPlayer/Components/PlayerTools/PlayerTools.tsx:6]: PlayerToolsPropsInterface

File: app/Components/SmallPlayer/Components/ProgressBar/ProgressBar.tsx [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+    if (props.progressRef.current) {
+      const progressPercentage =
+        (audioPlayer.currentTime / audioPlayer.duration) * 100;
+      props.progressRef.current.value = String(progressPercentage);
+    }
+  }, [audioPlayer.currentTime, audioPlayer.duration, props.progressRef]);
+
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    props.onChange(e);
+    const newTime = (e.target.valueAsNumber / 100) * audioPlayer.duration;
```
Comment (pailodze) [inline app/Components/SmallPlayer/Components/ProgressBar/ProgressBar.tsx:20]: newTime is not descirptive name

File: app/Components/SmallPlayer/Hooks/use-audio.hook.ts [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+
+export const useAudioPlayer = (songs: Song[]) => {
+  const [audioPlayer, setAudioPlayer] = useRecoilState(audioPlayerState);
+  const audioRef = useRef<HTMLAudioElement>(null);
+  const progressRef = useRef<HTMLInputElement>(null);
+  const ipadProgressRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    const handleTimeUpdate = () => {
+      if (audioRef.current) {
```
Comment (pailodze) [inline app/Components/SmallPlayer/Hooks/use-audio.hook.ts:14]: if (!audioRef.current) { return }

this code is called guard clause code

File: app/Components/SmallPlayer/Hooks/use-audio.hook.ts [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+  const handleProgressChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (audioRef.current) {
+      const newTime = (e.target.valueAsNumber / 100) * audioPlayer.duration;
+      audioRef.current.currentTime = newTime;
+      setAudioPlayer((prev) => ({ ...prev, currentTime: newTime }));
+    }
+  };
+
+  const playMusic = () => {
+    if (audioRef.current) {
```
Comment (pailodze) [inline app/Components/SmallPlayer/Hooks/use-audio.hook.ts:87]: same here

File: app/Components/SmallPlayer/Hooks/use-audio.hook.ts [hunk 26]
```diff
@@ -0,0 +1,168 @@
+import { useRef, useEffect } from 'react';
```
Comment (pailodze) [inline app/Components/SmallPlayer/Hooks/use-audio.hook.ts:1]: Hooks should be hooks

File: app/Components/SmallPlayer/Hooks/use-audio.hook.ts [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+import { useRecoilState } from 'recoil';
+import { audioPlayerState } from '@/app/Atoms/states';
+import { Song } from '@/app/Interfaces/Interfaces';
+
```
Comment (pailodze) [inline app/Components/SmallPlayer/Hooks/use-audio.hook.ts:5]: hook name useAudio.tsx

File: app/Components/SmallPlayer/SmallPlayer.tsx [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+    handleProgressChange,
+    playMusic,
+    handleTenSecondsBack,
+    handleNextSong,
+    handlePreviousSong,
+    handleVolumeDown,
+    handleVolumeUp,
+  } = useAudioPlayer(props.songs);
+
+  const isPlaying = audioRef.current ? !audioRef.current.paused : false;
```
Comment (pailodze) [inline app/Components/SmallPlayer/SmallPlayer.tsx:24]: extract it to useAudioPlayer

**Review by pailodze — CHANGES_REQUESTED** [review]

**Review by Mari-Codehub — APPROVED** [review]

**Pull request merged**